### PR TITLE
Fix git timeout handling in version resolution (#463)

### PR DESCRIPTION
**Pull request by hollisakins**
## Summary

Fix a critical bug where git timeouts during version string resolution could silently produce bare release strings (e.g., `0.5.1`) that bypass deploy's release validation. The fix distinguishes between "git answered no" (nonzero exit) and "git could not answer" (timeout/missing binary), marking any unresolved version with an `unresolved` local segment to ensure it always fails the release check.

## Key Changes

- **New `_GitUnavailable` exception**: Distinguishes timeouts and missing git binary from normal git command failures (nonzero exits)
- **Enhanced `_run_git()` error handling**: 
  - Raises `_GitUnavailable` on `TimeoutExpired` or `FileNotFoundError`
  - Returns `None` only for `CalledProcessError` (git answered "no")
  - Increased timeout from 5s to 30s (safe because result is cached per-process)
- **Refactored `_git_version()` logic**:
  - Tracks `unresolved` flag across all git queries
  - Any `_GitUnavailable` exception sets the flag and preserves partial results
  - Handles distance as `int | None` to distinguish "unknown" from "zero"
  - Validates `rev-list` output is numeric; garbage output marks version unresolved
  - Timed-out dirty check no longer stamps tree as clean
- **Updated `_build_pep440()`**: 
  - Accepts `distance: int | None` parameter
  - Includes HEAD sha in unresolved versions (pins code when distance/dirtiness unknown)
  - Appends `unresolved` to local segment when needed
- **Added logging**: Warns when version resolution is degraded due to git unavailability
- **Comprehensive test coverage**: New `TestGitUnavailable` class with 10 regression tests covering timeout scenarios, garbage output, and edge cases

## Notable Details

- Version strings from fully-answering git are byte-identical to before; no scientific output changes
- Resolution is `lru_cache`d per process, so the 30s timeout is a once-per-run cap
- All unresolved versions carry the `unresolved` local segment, ensuring they fail `deploy._is_release_version()` and trigger the warn-and-confirm path
- Handles cascading failures gracefully: if describe times out, the no-tag synthetic version is still marked unresolved

https://claude.ai/code/session_012rJaau51z3ia818ejE1ixA

### CODE REVIEW

File: pipeline/campfire_pipeline/common/version.py [hunk 1]
```diff
@@ -51,6 +73,20 @@ def _repo_root() -> Path:
...
+
+    Distinct from a nonzero exit (git *answering* "no", e.g. ``describe``
+    finding no matching tag), which ``_run_git`` reports as ``None``.
+    """
+
+
+# Generous because the result is cached per process, so this is a once-per-run
+# cap, and a timeout here permanently degrades the CMPFRVER provenance for
+# everything the process writes (#463: git on contended NFS blew a 5 s budget).
+_GIT_TIMEOUT = 30
```
[the remark this answers is not in the document]
Comment (hollisakins) [reply pipeline/campfire_pipeline/common/version.py]: Addressed in 3e10b3e: the 30 s allowance is now a single wall-clock deadline created at the top of `_git_version()` and passed to every probe. Once a hung git exhausts it, the remaining probes raise `_GitUnavailable` immediately without spawning a subprocess, so the worst-case startup stall is 30 s total rather than per probe. Covered by the new `TestSharedDeadline` tests.

---
_Generated by [Claude Code](https://claude.ai/code)_
